src/sp/it/pl/main/AppActions.java

```diff
@@ -396,29 +403,76 @@ public void openImageFullscreen(File image, Screen screen) {
 		else add1timeEventHandler(window, WindowEvent.WINDOW_SHOWN, t -> r.apply());
 	}
 
+	/**
+	 * The check whether file exists, is accessible or of correct type/format is left on the caller and behavior in
+	 * such cases is undefined.
+	 */
+	@Blocks
 	public void printAllImageFileMetadata(File file) {
+		throwIfFxThread();
+
+		String t = "Metadata of " + file.getPath();
 		try {
-			StringBuilder sb = new StringBuilder("Metadata of ").append(file.getPath());
+			StringBuilder sb = new StringBuilder();
 			com.drew.metadata.Metadata metadata = ImageMetadataReader.readMetadata(file);
 			metadata.getDirectories().forEach(d -> {
 				sb.append("\nName: ").append(d.getName());
 				d.getTags().forEach(tag -> sb.append("\n\t").append(tag.toString()));
 			});
-			APP.widgetManager.widgets.find(Widgets.LOGGER, WidgetSource.ANY); // open console automatically
-			System.out.println(sb.toString());
-		} catch (IOException|ImageProcessingException e) {
-			e.printStackTrace();    // TODO: improve
+			t = t + sb.toString();
+		} catch (IOException | ImageProcessingException e) {
+			t = t + "\n" + stackTraceAsString(e);
+		}
+		String text = t;
+		runFX(() -> APP.widgetManager.widgets.find(TextDisplayFeature.class, NEW).ifPresent(w -> w.showText(text)));
```

Comment:
You mean why does the method not return the text and leave the caller to do whatever with it? Because nobody needs such method. What we do need is an action that does that, so the method is unnecessary to be standalone and is inlined in the action.

I notice this mistake lots of people repeat (I certainly did) - refactor everything into a method, just in case. Remove all duplication, always. Make everything completely generic. With time this leads to downward spiral of maintenance pain and this project hurts all over.

Instead of coupling, look at it as encapsulation. Making it a method will complicate API with method nobody needs or understands and add overheads we never wanted. What would the method do? Why does anyone need audio metadata as one giant String formatted in some specific way?. It is useless. And forget formatting, that thing uses toStirng() on everything!

Making this a method would be like releasing an undocumented utility jar on an AppStore.

If you provide a performant (StringBuffer) and consistent (in output format) implementations, I wont mind, although I would still question why one would need a String version of file metadata. I warn you though, the only way to do this is to provide your own toString converter for every audio Tag and TagField implementation in the jaudiotagger library and then maintain it for all eternity. For our purposes this is an overkill, but if you have time, go for it.